Add WAVEWATCH III as new coupled wave component

sbrus89 opened the pull request:
This is a stealth feature PR that adds the WAVEWATCH III submodule, cime configuration, namelist generation, build infrastracture, and coupling interface. 

Wave only, G case + waves, and B case + waves compsets are included along with model resolutions that include the wave mesh intended for v1 and v2 LR configurations. 

The initial coupled, active wave feedback is a parameterization for Langmuir mixing. Details can be found in the overview document below:

New Feature Overview Document: [E3SM_New_Feature_Overview___Waves.pdf](https://github.com/E3SM-Project/E3SM/files/8980373/E3SM_New_Feature_Overview___Waves.pdf)

MPAS-Analysis and E3SM-diags results for 10 year B-Case comparison:
[Coupled waves vs. Observations](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/sbrus/ww3_merge/waves)
[Reference case vs. Observations](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/sbrus/ww3_merge/ref)
[Coupled waves vs. Reference case](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/sbrus/ww3_merge/compare)

PACE performance results:
[v1 LR on Chrysalis](https://pace.ornl.gov/exp-details/95598)
[v2 LR resolution on Anvil](https://pace.ornl.gov/exp-details/108664)

[NML] for configurations with active ocean
[BFB]

sbrus89 commented:
Looks like the CI failed because I forgot to push an extra commit to https://github.com/E3SM-Project/WW3. It should be fixed now.

sarats commented:
Minor:
I looked at the PACE data and wondered if we can instrument the important important sub-processes (computationally time-consuming preferably). Right now, the entire time in under `CPL:WAV_RUN`. Can your add other timers to your driver/wrapper function if possible?

https://pace.ornl.gov/flamegraph/108664/1728/

rljacob commented:
A test should be added for this.  If the only options are expensive tests, we can put it in the extra-coverage suite.

sbrus89 commented:
I'm back from parental leave now, so I'll be able to better engage with this now.

sbrus89 commented:
I have rebased to resolve conflicts and checked that it is BFB with the initial submission.

sbrus89 commented:
@sarats, I agree that better timing granularity would be useful. Would it be possible for this be part of a future PR? 

I think this would require either bringing the E3SM timing routines into WW3 or pulling out the calls from the WW3 timestep driver (`w3wave`) inside `WAV_RUN_MCT`.  I need to think more about which would be a better direction to go with.

sbrus89 commented:
@rljacob, Thanks for pointing that out. I added a smoke test in the extended coverage suite.

sbrus89 commented:
I'm not sure about what is causing the CI to fail. It looks like an issue with `singularity`?

sbrus89 commented:
Thanks for helping sort out the CI issue, @rljacob!

mark-petersen commented:
@sbrus89 just to make sure I understand, WW3 is never run with stand-alone MPAS-Ocean, only when coupled within E3SM, right?  Can WW3 also run stand-alone within the E3SM directory structure, or is that not the intent?

sbrus89 commented on components/mpas-ocean/src/shared/Makefile:
```diff
@@ -226,11 +227,15 @@ mpas_ocn_vel_pressure_grad.o: mpas_ocn_constants.o
 mpas_ocn_vertical_remap.o: mpas_ocn_config.o mpas_ocn_constants.o mpas_ocn_diagnostics_variables.o mpas_ocn_mesh.o mpas_ocn_vertical_regrid.o
+=======
+mpas_ocn_stokes_drift.o: mpas_ocn_constants.o mpas_ocn_config.o mpas_ocn_diagnostics_variables.o mpas_ocn_mesh.o
+>>>>>>> c50e0762f4... Subtree merge of Stokes drift MPAS-Ocean changes
```
Thanks for catching that @mark-petersen, it's fixed now.

sbrus89 commented:
@mark-petersen, that's correct. WW3 is only intended to run within E3SM and not with stand-alone MPAS-Ocean. Stand alone WW3 can be complied from within the WW3 submodule, but no stand-alone executable is generated as a part of the E3SM build. There are compsets to run data-forced WW3, which is essentially the same as stand-alone.

mark-petersen commented on components/mpas-ocean/src/Registry.xml:
```diff
@@ -655,6 +659,20 @@
+	</nml_record>
+	<nml_record name="wave_coupling" mode="forward">
+		<nml_option name="config_use_active_wave" type="logical" default_value=".false." units="NA"
+					description="Flag for using prognostic waves"
```
Does this flag turn on coupling with WW3? Could you add more to the description here: 
1. is this required to be on with WW3 coupling in E3SM
2. What happens if you turn this on in stand-alone, or in E3SM without WW3?

It seems like the master switch for WW3 should be in E3SM, not here. But perhaps this just says whether to apply the coupling fields in MPAS-Ocean.

sbrus89 replied:
@mark-petersen, thanks for bringing this up; I wanted to make sure I got feedback about this. This isn't required for WW3 coupling in E3SM. Right now this flag only controls the MPAS-O package for waves, so that none of the waves related arrays are allocated when WW3 isn't being used.

mark-petersen replied:
In that case, I would name it `config_allocate_wave_arrays`. But it must also control the use of those arrays within MPAS-Ocean? Or the planned use of them in the future? Otherwise, what would the arrays be for?

sbrus89 replied:
I forgot to mention that right now this also controls the calculation of the stokes drift profiles as well. Another thing to mention is that I've been using MPAS streams as the primary means of outputting wave variables because 1) the WW3 output is really cumbersome, and 2) it makes it possible to use MPAS-Analysis for waves. So this flag also facilitates that streams output in absence of coupled parameterizations.

mark-petersen commented on components/mpas-ocean/src/Registry.xml:
```diff
@@ -655,6 +659,20 @@
+					description="Flag for using prognostic waves"
+					possible_values=".true. or .false."
+		/>
+		<nml_option name="config_n_stokes_drift_wavenumber_partitions" type="integer" default_value="6" units="NA"
```
Is there an option to turn stokes drift on or off? Or is that in the `config_use_active_wave`? If so, please add that to the description.

sbrus89 replied:
`config_use_active_wave` turns off/on all wave related fields and  the stokes drift profile calculation. `config_cvmix_kpp_use_active_wave` specifically turns on/off the active wave Langmuir parameterization.

sbrus89 replied:
I should probably add a condition that ensures `config_cvmix_kpp_use_active_wave` can't be `.true.` if `config_use_active_wave ` is `.false.`

mark-petersen replied:
Yes, good idea. OK, I think the flag name `config_use_active_wave` is fine if you add more details to the description (i.e. turns on all wave related fields and the stokes drift profile calculation) and also say that it requires to be run in E3SM with WW3 on for active_wave to work.

sbrus89 replied:
Great, I'll be sure to do that.

sbrus89 replied:
I tried to be more descriptive in the latest commit, let me know what you think.

mark-petersen commented:
With default flags, passes nightly suite with gnu debug and intel debug (with Makefile conflict fixed).

@sbrus89 what flags should I change to exercise the new code in stand-alone MPAS-Ocean?

sbrus89 commented:
> @sbrus89 what flags should I change to exercise the new code in stand-alone MPAS-Ocean?

I don't think there is a good way to exercise the new code in stand-alone, since it requires wave fields from the coupler. How is this handled for other components external to the ocean core?

mark-petersen commented:
If I set `config_use_active_wave=.true.` and `config_cvmix_kpp_use_active_wave=.true.` then won't it just run in stand-alone with zeros in the wave arrays? I just tried it with QU240 performance tests and there were no errors. I'll try it in E3SM next.

sbrus89 commented:
That's true, good point.

mark-petersen commented:
@sbrus89, I am testing in E3SM with some automated scripts. I see in your pdf the new comp sets and resolutions. Do you think these make sense together? on chrysalis:
```
./create_test SMS_D.CFSR_wQU225EC60to30.VWW3-IAF.chrysalis_gnu -q acme-small --walltime 30:00
./create_test SMS_D.T62_oEC60to30v3_wQU225EC60to30.GMPAS-JRA-WW3.chrysalis_gnu -q acme-small --walltime 30:00
```
and same ones on cori:
```
./create_test SMS_D.CFSR_wQU225EC60to30.VWW3-IAF.cori-haswell_gnu -q debug --walltime 00:30:00 #####
./create_test SMS_D.T62_oEC60to30v3_wQU225EC60to30.GMPAS-JRA-WW3.cori-haswell_intel -q debug --walltime 00:30:00
```
We can run partition (PEM) and thread (PET) tests on those. Are there other combinations of resolutions and comp sets that make sense? In particular, ones that fit in a debug queue? A small number of time steps is fine.

sbrus89 commented:
@mark-petersen, yes I think those make sense to run for testing.

darincomeau commented:
hmm `T62_oEC60to30v3_wQU225EC60to30.GMPAS-JRA-WW3` doesn't look correct; JRA uses `TL319`. I'm checkout out this branch now to take a closer look.

sbrus89 commented:
@darincomeau, that's correct. It must be a typo in the pdf. It is `TL319_EC30to60E2r2_wQU225EC30to60E2r2` is in `config_grids.xml`

sbrus89 commented:
so @mark-petersen, `./create_test SMS_D.TL319_EC30to60E2r2_wQU225EC30to60E2r2.GMPAS-JRA-WW3.chrysalis_gnu` would be the one to run.

sbrus89 commented:
@mark-petersen, I've been running into issues with `create_test` that have to do with archiving at the end of a run. I may need some help sorting that out.

sbrus89 commented:
@darincomeau, feel free to add yourself as a reviewer. No pressure though.

darincomeau commented:
I'm happy enough to just offer some comments from the peanut gallery.

@mark-petersen if you've done that test with `GMPAS-JRA-WWW` can you point me to the output? I just want to check the `datm` streams to make sure it's pulling the correct files (and git is being *very* slow for me on LCRC).

darincomeau commented:
I found the `GMPAS-JRA-WW3` compset and made a comment there. I'd also recommend changing the compset name to `GMPAS-JRA1p4-WW3`.

mark-petersen commented:
Results of tests so far:
```
./create_test SMS_D.CFSR_wQU225EC60to30.VWW3-IAF.chrysalis_gnu -q acme-small --walltime 30:00
./create_test SMS_D.CFSR_wQU225EC60to30.VWW3-IAF.cori-haswell_gnu -q debug --walltime 00:30:00
paths:
/lcrc/group/e3sm/ac.mpetersen/scratch/chrys/SMS_D.CFSR_wQU225EC60to30.VWW3-IAF.chrysalis_gnu.20220830_163213_2rcjdn/run
/global/cscratch1/sd/mpeterse/e3sm_scratch/cori-haswell/SMS_D.CFSR_wQU225EC60to30.VWW3-IAF.cori-haswell_gnu.20220830_144334_uawuqy/run
```
These initialize ww3, as seen in the `e3sm.log.*` file:
```
 194 0:                    *** WAVEWATCH III Grid preprocessor ***
 195 0:                ===============================================
 196 0:
 197 0:   Grid name : wQU225EC60to30
...
```

<details>
<summary> but end with a seg fault</summary> <p>

```
1867 0:  Filling 3D look-up table for SIN4. please wait
1868 0:  WAVEWATCH III ST4 TABLE FOR STRESS 2018-06-08   6.05416155       9.49999969E-03   1.42999995             100         200          50   6.00000005E-03  0.400000006       9.80599976
1869 0:
1870 0:   Summary grid statistics :
1871 0:  --------------------------------------------------
1872 0:        Number of longitudes      :         0
1873 0:        Number of latitudes       :         0
1874 0:        Number of grid points     :     56131
1875 0:        Number of sea points      :     56131 (100%)
1876 0:        Number of input b. points :         0
1877 0:        Number of land points     :         0
1878 0:        Number of excluded points :         0
1879 0:
1880 0:
1881 0:   End of program
1882 0:  ========================================
1883 0:          WAVEWATCH III Grid preprocessor
1884 0:
1885 0:  SI nbPlus=      323034  nbMinus=           0  nbZero=      238276
1886 0:
1887 0: Program received signal SIGSEGV: Segmentation fault - invalid memory reference.
1888 0:
1889 0: Backtrace for this error:
1890 0: #0  0x15554f18d3ff in ???
1891 0: #1  0x858713 in __w3iorsmd_MOD_w3iors
1892 0:  at /gpfs/fs1/home/ac.mpetersen/repos/E3SM/pr/components/ww3/src/WW3/model/tmp/w3iorsmd.F90:384
1893 0: #2  0x7ae161 in __w3initmd_MOD_w3init
1894 0:  at /gpfs/fs1/home/ac.mpetersen/repos/E3SM/pr/components/ww3/src/WW3/model/tmp/w3initmd.F90:1250
1895 0: #3  0x615122 in __wav_comp_mct_MOD_wav_init_mct
1896 0:  at /gpfs/fs1/home/ac.mpetersen/repos/E3SM/pr/components/ww3/src/cpl/wav_comp_mct.F90:820
1897 0: #4  0x441f12 in __component_mod_MOD_component_init_cc
1898 0:  at /gpfs/fs1/home/ac.mpetersen/repos/E3SM/pr/driver-mct/main/component_mod.F90:248
1899 0: #5  0x4298ab in __cime_comp_mod_MOD_cime_init
1900 0:  at /gpfs/fs1/home/ac.mpetersen/repos/E3SM/pr/driver-mct/main/cime_comp_mod.F90:1476
1901 0: #6  0x43ad74 in cime_driver
1902 0:  at /gpfs/fs1/home/ac.mpetersen/repos/E3SM/pr/driver-mct/main/cime_driver.F90:122
1903 0: #7  0x43aeb4 in main
1904 0:  at /gpfs/fs1/home/ac.mpetersen/repos/E3SM/pr/driver-mct/main/cime_driver.F90:23
1905 srun: error: chr-0493: task 0: Segmentation fault (core dumped)
1906 srun: Terminating job step 216608.0
```
</details>
and the file wav.log.* has a large amount of output. 

The second set do not work at all. On chrysalis the build hangs, and on cori it compiles but dies on init with `out-of-memory` error.
```
./create_test SMS_D.T62_oEC60to30v3_wQU225EC60to30.GMPAS-JRA-WW3.chrysalis_gnu -q acme-small --walltime 30:00
./create_test SMS_D.T62_oEC60to30v3_wQU225EC60to30.GMPAS-JRA-WW3.cori-haswell_intel -q debug --walltime 00:30:00
```
But @sbrus89 and @darincomeau point out that it should use TL319.

@sbrus89, I would like to test all seven comp sets with smoke tests, with the lowest resolution available - hopefully to fit in the debug queue time for 5-10 time steps. Could you test those and provide them here? We could start with the simplest and work through the errors. Thanks!

mark-petersen commented:
I tried this one, with `CFS` in both resolution and comp set:
```
./create_test SMS_D_Ln3.CFSv2_wQU225EC60to30.VWW3-CFSv2.cori-haswell_gnu -q debug --walltime 30:00
```
and it dies trying to access the files:
```
        wget failed with output:  and errput --2022-08-31 06:08:29--  https://web.lcrc.anl.gov/public/e3sm/inputdata/wav/cfsv2/domain.ocn.CFSv2_CFSv2.200616.nc
        Resolving web.lcrc.anl.gov (web.lcrc.anl.gov)... 140.221.70.30
        Connecting to web.lcrc.anl.gov (web.lcrc.anl.gov)|140.221.70.30|:443... connected.
        HTTP request sent, awaiting response... 403 Forbidden
        2022-08-31 06:08:29 ERROR 403: Forbidden.

        wget failed with output:  and errput --2022-08-31 06:08:29--  https://web.lcrc.anl.gov/public/e3sm/inputdata/wav/cfsv2/sandy_wnd10m_ww3.nc
        Resolving web.lcrc.anl.gov (web.lcrc.anl.gov)... 140.221.70.30
        Connecting to web.lcrc.anl.gov (web.lcrc.anl.gov)|140.221.70.30|:443... connected.
        HTTP request sent, awaiting response... 403 Forbidden
        2022-08-31 06:08:30 ERROR 403: Forbidden.
```

If you go to these files here: https://web.lcrc.anl.gov/public/e3sm/inputdata/wav/cfsv2/ and click on the files, you can see that the permissions are not set to be readable on the server.

mark-petersen commented:
Also tried matching CFSR in comp set and resolution:
```
 ./create_test SMS_D_Ln3.CFSR_wQU225EC60to30.VWW3-CFSR.chrysalis_gnu -q acme-small --walltime 30:00
```
dies with out of bounds on chrysalis:
```
vi /lcrc/group/e3sm/ac.mpetersen/scratch/chrys/SMS_D_Ln3.CFSR_wQU225EC60to30.VWW3-CFSR.chrysalis_gnu.20220831_080432_mf8vsp/run/e3sm.log.216850.220831-081328

109 0: At line 626 of file /gpfs/fs1/home/ac.mpetersen/repos/E3SM/pr/externals/mct/mpeu/m_List.F90
110 0: Fortran runtime error: Index '1' of dimension 2 of array 'alist%lc' above upper bound of 0
```
and same one on cori:
```
./create_test SMS_D_Ln3.CFSR_wQU225EC60to30.VWW3-CFSR.cori-haswell_intel -q debug --walltime 00:30:00
```

<details>
<summary> dies with divide by zero: </summary> <p>

```
vi /global/cscratch1/sd/mpeterse/e3sm_scratch/cori-haswell/SMS_D_Ln3.CFSR_wQU225EC60to30.VWW3-CFSR.cori-haswell_intel.20220831_060509_xedjxj/run/e3sm.log.62354083.220831-061208

1954 0:  WAVEWATCH III ST4 TABLE FOR STRESS 2018-06-08   6.054162      9.4999997E-03
1955 0:    1.430000             100         200          50  6.0000001E-03
1956 0:   0.4000000       9.806000
1957 0:
1958 0:   Summary grid statistics :
1959 0:  --------------------------------------------------
1960 0:        Number of longitudes      :         0
1961 0:        Number of latitudes       :         0
1962 0:        Number of grid points     :     56131
1963 0:        Number of sea points      :     56131 (100%)
1964 0:        Number of input b. points :         0
1965 0:        Number of land points     :         0
1966 0:        Number of excluded points :         0
1967 0:
1968 0:
1969 0:   End of program
1970 0:  ========================================
1971 0:          WAVEWATCH III Grid preprocessor
1972 0:
1973 0:  SI nbPlus=      323034  nbMinus=           0  nbZero=      238276
1974 0: forrtl: error (73): floating divide by zero
1975 0: Image              PC                Routine            Line        Source
1976 0: e3sm.exe           0000000002543254  Unknown               Unknown  Unknown
1977 0: e3sm.exe           0000000002162420  Unknown               Unknown  Unknown
1978 0: e3sm.exe           00000000025FBB27  Unknown               Unknown  Unknown
1979 0: e3sm.exe           00000000010E53F5  w3src4md_mp_w3spr         220  w3src4md.F90
1980 0: e3sm.exe           00000000014F6447  w3srcemd_mp_w3src         557  w3srcemd.F90
1981 0: e3sm.exe           00000000011EBBD2  w3wavemd_mp_w3wav        1287  w3wavemd.F90
1982 0: e3sm.exe           000000000092AB2E  wav_comp_mct_mp_w        1208  wav_comp_mct.F90
1983 0: e3sm.exe           000000000047C926  component_mod_mp_         751  component_mod.F90
1984 0: e3sm.exe           000000000043A183  cime_comp_mod_mp_        2928  cime_comp_mod.F90
1985 0: e3sm.exe           0000000000464288  MAIN__                    153  cime_driver.F90
```
</details>

Maybe I'm choosing the wrong combinations? @sbrus89 I'll stop testing, and let you provide further guidance here. Thanks.

rljacob commented on cime_config/config_archive.xml:
```diff
@@ -141,4 +141,21 @@
+  <comp_archive_spec compname="ww3" compclass="wav">
+    <rest_file_extension>restart.ww3</rest_file_extension>
+    <hist_file_extension>out_pnt.ww3</hist_file_extension>
+    <hist_file_extension>out_grd.ww3</hist_file_extension>
```
I don't think you need "ww3" here.  It expects "compname" to be in file name.

sbrus89 commented:
@mark-petersen, I believe I have fixed the issues with `SMS_D_Ln3.CFSR_wQU225EC60to30.VWW3-CFSR` and `SMS_D_Ln3.CFSv2_wQU225EC60to30.VWW3-CFSv2`. One issue I ran across is that it appears that the datm encounters an out of bounds error when no field from the `stifld` list from https://github.com/E3SM-Project/E3SM/blob/bbcb4490f3fa1e4869527666d25bd835196aa54d/components/data_comps/datm/src/datm_comp_mod.F90#L180-L193
is specified. Only the u and v wind velocity components are needed for these cases, but I added a dummy field for tbot as a work-around.

sbrus89 commented:
Also, for the short term achieve step to pass, the lines: https://github.com/ESMCI/cime/blob/98106c908bd3b5514239e56efdc7ab6a53524065/CIME/case/case_st_archive.py#L1187-L1188
need to be removed. This would require a cime PR and submodule update in E3SM.

mark-petersen commented:
Thanks @sbrus89, I know adding a new component is a lengthy process.

Testing again and gathering results. These have build errors on cori:
```
./create_test SMS_D_Ln3.CFSv2_wQU225EC60to30.VWW3-CFSv2.cori-haswell_gnu -q debug --walltime 30:00
./create_test SMS_D_Ln3.CFSv2_wQU225EC60to30.VWW3-CFSv2.cori-haswell_intel -q debug --walltime 30:00
```
these build but have run-time errors in chrysalis:
```
./create_test SMS_D_Ln3.CFSv2_wQU225EC60to30.VWW3-CFSv2.chrysalis_gnu -q acme-small --walltime 30:00
./create_test SMS_D_Ln3.CFSv2_wQU225EC60to30.VWW3-CFSv2.chrysalis_intel -q acme-small --walltime 30:00
```
They die with a cime run-time error `LVD not found, all data is after yearLast` in line 114 of shr_abort_mod.F90

This one builds and dies within the run:
```
./create_test SMS_D_Ln3.CFSR_wQU225EC60to30.VWW3-CFSR.cori-haswell_intel -q debug --walltime 00:30:00
```
see directory
```
/global/cscratch1/sd/mpeterse/e3sm_scratch/cori-haswell/SMS_D_Ln3.CFSR_wQU225EC60to30.VWW3-CFSR.cori-haswell_intel.20220908_110259_ze7y1a/run

vi e3sm.log.62617497.220908-114918

PIO: FATAL ERROR: Aborting... An error occured, Reading variable (, varid=8) from file (/global/cfs/cdirs/e3sm/inputdata/wav/cfsr/wnd10mx0.5.gdas.200506.ww3.nc, ncid=16) failed. Inquiring variable data type failed. NetCDF: Variable not found (err=-49). Aborting since the error handler was set to PIO_INTERNAL_ERROR... (/global/u2/m/mpeterse/repos/E3SM/pr/externals/scorpio/src/clib/pio_darray.c: 2501)

0: Rank 0 [Thu Sep  8 11:49:31 2022] [c6-5c0s10n2] application called MPI_Abort(MPI_COMM_WORLD, -1) - process 0
0: forrtl: error (76): Abort trap signal

0: e3sm.exe           0000000001CABF48  piodarray_mp_read         421  piodarray.F90.in
0: e3sm.exe           0000000001CA3487  piodarray_mp_read         393  piodarray.F90.in
0: e3sm.exe           000000000175AB26  shr_dmodel_mod_mp         984  shr_dmodel_mod.F90
0: e3sm.exe           00000000017535DB  shr_dmodel_mod_mp         716  shr_dmodel_mod.F90
0: e3sm.exe           000000000191508A  shr_strdata_mod_m         891  shr_strdata_mod.F90
0: e3sm.exe           00000000008AF491  datm_comp_mod_mp_         664  datm_comp_mod.F90
0: e3sm.exe           00000000008AEADD  datm_comp_mod_mp_         551  datm_comp_mod.F90
0: e3sm.exe           00000000008A4F06  atm_comp_mct_mp_a         167  atm_comp_mct.F90
0: e3sm.exe           000000000046D682  component_mod_mp_         257  component_mod.F90
0: e3sm.exe           0000000000415A89  cime_comp_mod_mp_        1440  cime_comp_mod.F90
0: e3sm.exe           0000000000464121  MAIN__                    122  cime_driver.F90
```

sbrus89 commented:
@mark-petersen, thanks for re-testing. I'll take a look at the errors.

sbrus89 commented:
@mark-petersen, you asked for this ealier, sorry I didn't follow up. The main ones we should focus on testing are:
 - `CFSv2_wQU225EC60to30.VWW3-CFSv2` : wave only, CFSv2 forcing
 - `CFSR_wQU225EC60to30.VWW3-CFSR` : wave only, CFSR forcing
 - `T62_oEC60to30v3_wQU225EC60to30.GMPAS-IAF-WW3` : COREII forced G-case + waves (v1 res)
 - `TL319_EC30to60E2r2_wQU225EC30to60E2r2.GMPAS-JRA-WW3` : JRA forced G-case + waves (v2 res)
 - `ne30_oECv3_wQU225EC60to30.WCYCLSSP585-WW3` : future scenerio B-case + waves (v1 res)
 - `ne30pg2_EC30to60E2r2_wQU225EC30to60E2r2.WCYCL1850-WW3` : piControl B-case + waves (v2 res)

sbrus89 commented:
## CFSv2_wQU225EC60to30.VWW3-CFSv2

- Chrysalis
  - gnu debug: PASS
  - intel debug: PASS
  - gnu: PASS
  - intel: PASS
- Cori
  - gnu debug: PASS
  - intel debug: FAIL (divide by zero in w3src4md)
  - gnu: PASS
  - intel: PASS

## CFSR_wQU225EC60to30.VWW3-CFSR

- Chrysalis
  - gnu debug: PASS
  - intel debug: PASS
  - gnu: PASS
  - intel: PASS
- Cori
  - gnu debug: PASS
  - intel debug: FAIL (divide by zero in w3src4md)
  - gnu: PASS
  - intel: PASS

## T62_oEC60to30v3_wQU225EC60to30.GMPAS-IAF-WW3

- Chrysalis
  - gnu debug: FAIL (unknown error)
  - intel debug: PASS
  - gnu: PASS
  - intel: PASS
- Cori
  - Tests FAIL: (Out of memory, need PE layout)

## TL319_EC30to60E2r2_wQU225EC30to60E2r2.GMPAS-JRA-WW3

- Chrysalis
  - gnu debug: PASS
  - intel debug: PASS
  - gnu: PASS
  - intel: PASS
- Cori
  - Tests FAIL: (Out of memory, need PE layout)

## ne30_oECv3_wQU225EC60to30.WCYCLSSP585-WW3

- Chrysalis
  - Tests FAIL: (ELM compatibility)
- Cori
  - Tests FAIL: (ELM compatibility)

## ne30pg2_EC30to60E2r2_wQU225EC30to60E2r2.WCYCL1850-WW3

- Chrysalis
  - gnu debug: PASS
  - intel debug: PASS
  - gnu: PASS
  - intel: PASS
- Cori
  - Tests FAIL: (Out of memory, need PE layout)

sbrus89 commented:
@mark-petersen, the comment above summarizes the results of my current SMS test results. 

Everything PASSES on Chrysalis except T62_oEC60to30v3_wQU225EC60to30.GMPAS-IAF-WW3 with GNU debug (I've been trying to track down the issue inside WW3, intel debug passes) and ne30_oECv3_wQU225EC60to30.WCYCLSSP585-WW3, which has an ELM issue related to v1/v2 compatibility (I had previously used this compset with v1 restart files). 

On Cori, the larger tests don't run because I didn't create PE layouts for them. Also, there's an error inside WW3 for intel debug on the CFSv2 and CFSR cases (gnu debug passes)

rljacob commented:
I think it is ok to merge this.  Remaining problems can have issues opened for them.

You shouldn't spend any time on Cori since it is going away less than 4 months from now.  Gnu is not a compiler we use in production on Chrysalis so that can wait.

sbrus89 commented:
I just want to mention that, currently, WW3-enabled cases cannot be built in parallel. This is due to a legacy preprocessing script that must be run on the WW3 source code to produce compilable code for a given set of model options. A race condition occurs when this script is run in parallel. 

Future versions of WW3 will not require this preprocessing step. An E3SM PR will be submitted to update the WW3 submodule build process once a new WW3 version is released.

mark-petersen reviewed (approved):
Passes build and run for:
```
./create_test SMS_D_Ln3.CFSv2_wQU225EC60to30.VWW3-CFSv2.chrysalis_gnu -q acme-small --walltime 30:00
./create_test SMS_D_Ln3.CFSR_wQU225EC60to30.VWW3-CFSR.chrysalis_gnu -q acme-small --walltime 30:00
./create_test SMS_D_Ln3.CFSv2_wQU225EC60to30.VWW3-CFSv2.chrysalis_intel -q acme-small --walltime 30:00 
(run did not complete in allocated time, but runs otherwise)
./create_test SMS_D_Ln3.CFSR_wQU225EC60to30.VWW3-CFSR.chrysalis_intel -q acme-small --walltime 30:00 
(run did not complete in allocated time, but runs otherwise)
./create_test SMS_Ln3.CFSv2_wQU225EC60to30.VWW3-CFSv2.cori-haswell_gnu -q debug --walltime 30:00 
(that one also passes debug compile and run, but runs out of time in the debug queue)
```
Thanks again @sbrus89 for your hard work here. My review was for functionality and testing, not for quality of simulation. But I think the simulation quality was tested by Steve and others, so I'm happy for this to be merged now.

mark-petersen commented on components/mpas-seaice/driver/ice_comp_mct.F:
```diff
@@ -2544,6 +2544,7 @@ subroutine ice_export_mct(i2x_i, errorCode)   !{{{
             i2x_i % rAttr(index_i2x_Si_tref  ,n)  = atmosReferenceTemperature2m(i)
             i2x_i % rAttr(index_i2x_Si_qref  ,n)  = atmosReferenceHumidity2m(i)
             i2x_i % rAttr(index_i2x_Si_snowh ,n)  = snowVolumeCell(i) / ailohi
+            i2x_i % rAttr(index_i2x_Si_ithick,n)  = iceVolumeCell(i) / ailohi
```
@akturner please review these additions to the sea ice coupler variables, which were added for Wave Watch III.

mark-petersen commented on components/mpas-seaice/driver/mpassi_cpl_indices.F:
```diff
@@ -146,6 +147,7 @@ subroutine mpassi_cpl_indices_set( )
     index_i2x_Si_tref       = mct_avect_indexra(i2x,'Si_tref')
     index_i2x_Si_qref       = mct_avect_indexra(i2x,'Si_qref')
     index_i2x_Si_ifrac      = mct_avect_indexra(i2x,'Si_ifrac')
+    index_i2x_Si_ithick     = mct_avect_indexra(i2x,'Si_ithick')
```
@akturner, also here.

sbrus89 commented:
@jonbob, Just letting you know that all reviewers have approved this one.

jonbob commented:
Thanks for the reminder @sbrus89 -- it's on my list and I'll get it in as soon as possible

mark-petersen commented on components/mpas-ocean/cime_config/buildnml:
```diff
@@ -1314,6 +1314,7 @@ def buildnml(case, caseroot, compname):
             lines.append('    <var name="peakWaveFrequency"/>')
             lines.append('    <var name="peakWaveDirection"/>')
             lines.append('    <var name="significantWaveHeight"/>')
+            lines.append('    <var name=""langmuirNumber/>')
```
double quote out of place

sbrus89 replied:
thanks for catching that, @mark-petersen

darincomeau reviewed (changes requested):
@sbrus89 - this came up on today's mpas-devops call, and I hadn't realized that my previous code comments weren't showing up because they were grouped together as a 'review' that I never finished. Sorry about the lapse!

I had a couple minor change requests: 1) make units CF compliant, and 2) change JRA compset to JRA1p4-2018. The changes should all be commented on the code.

sbrus89 commented on components/mpas-ocean/src/Registry.xml:
```diff
@@ -3372,6 +3392,47 @@
 		/>
 
+                <!-- Waves coupling fields -->
+		<var name="stokesDriftZonal" type="real" dimensions="nVertLevels nCells Time" units="m s^{-1}"
```
@darincomeau, thanks for pointing this out. I'll make those changes.

sbrus89 commented on components/mpas-ocean/src/Registry.xml:
```diff
@@ -615,6 +619,20 @@
 					description="Option of Langmuir enhanced entrainment parameterization"
 					possible_values="NONE, LWF16, LF17, RWHGK16"
 		/>
+		<nml_option name="config_cvmix_kpp_use_active_wave" type="logical" default_value=".false." units="NA"
```
Thanks for mentioning this!

sbrus89 commented on components/ww3/cime_config/config_compsets.xml:
```diff
@@ -0,0 +1,47 @@
+
+  <compset>
+    <alias>GMPAS-JRA-WW3</alias>
+    <lname>2000_DATM%JRA_SLND_MPASSI_MPASO%DATMFORCED_DROF%JRA_SGLC_WW3</lname>
```
Great, I'll fix this.

sbrus89 commented on components/data_comps/datm/cime_config/namelist_definition_datm.xml:
```diff
@@ -2759,6 +2779,8 @@
       <value stream="presaero.cesm2_omip" datm_mode="CORE_IAF_JRA.*2018">1652</value> <!-- 1958-(6-1)*(2018-1958+1)-1 -->
       <value stream="topo.observed">1</value>
       <value stream="topo.cplhist">$DATM_CPLHIST_YR_START</value>
+      <value stream="CFSv2.UV">1</value>
```
@darincomeau, can you explain a little more about the change here? The CFSv2 and CFSR datm's were added for wave-only testing purposes and should be independent of `CORE_IAF_JRA`. I'm probably misunderstanding something though.

sbrus89 replied:
No problem!

darincomeau commented on components/mpas-ocean/src/Registry.xml:
```diff
@@ -3372,6 +3392,47 @@
+			 description="Frequency of peak of wave spectrum"
+			 packages="activeWavePKG"
+		/>
+		<var name="peakWaveDirection" type="real" dimensions="nCells Time" units="deg"
```
@sbrus89 yes, for direction I agree it should be `degree`. I hastily assumed it was temperature - thanks for checking that.

sbrus89 replied:
Perfect. I'll change that.

sbrus89 commented:
@darincomeau, let me know if the changes from the last two commits address your comments.

darincomeau commented on components/ww3/cime_config/config_compsets.xml:
```diff
@@ -30,8 +30,8 @@
-    <alias>GMPAS-JRA-WW3</alias>
-    <lname>2000_DATM%JRA_SLND_MPASSI_MPASO%DATMFORCED_DROF%JRA_SGLC_WW3</lname>
+    <alias>GMPAS-JRA-1p4-2018-WW3</alias>
+    <lname>2000_DATM%JRA-1p4-2018_SLND_MPASSI_MPASO%DATMFORCED_DROF%JRA_SGLC_WW3</lname>
```
You need to designate the `JRA-1p4-2018` for the `DROF` component as well as `DATM`:

```
<lname>2000_DATM%JRA-1p4-2018_SLND_MPASSI_MPASO%DATMFORCED_DROF%JRA-1p4-2018_SGLC_WW3</lname>
```

You don't necessarily need the `2018` part in the alias, so `GMPAS-JRA1p4-WW3` might be easier (but it's up to you). That's what we did for G-cases:
https://github.com/E3SM-Project/E3SM/blob/2491723ccacd2e960502789fb5c3f90182ec462d/components/mpas-ocean/cime_config/config_compsets.xml#L56

sbrus89 replied:
Very helpful, thanks. I overlooked the JRA for DROF.

darincomeau commented:
Thanks @sbrus89 for addressing these comments (and sorry my comments were sitting here hidden for weeks!).

sbrus89 commented:
No problem, @darincomeau. I've had that happen to me too. Thanks again for looking things over; I really appreciate the extra set of eyes.

jonbob commented:
after the most recent commit, testing of e3sm_developer on chrysalis shows only expected NML and FIELDLIST changes

merged to next

sbrus89 commented:
Thanks @jonbob!

jonbob commented:
merged to master and expected DIFFs and NML DIFFs blessed

The DIFFs are mostly due to new fields being passed through the coupler, which causes FIELDLIST DIFFs but no changes to answers. X-cases will change answers for the same reason, but have real DIFFs because the number of fields being passed changes and that impacts the actual values that configuration sends through the coupler
